lib/pengine/pe_digest.c

```diff
@@ -336,20 +337,20 @@ pe__calculate_digests(pe_resource_t *rsc, const char *task, guint *interval_ms,
  * \internal
  * \brief Calculate action digests and store in node's digest cache
  *
- * \param[in] rsc          Resource that action was for
- * \param[in] task         Name of action performed
- * \param[in] interval_ms  Action's interval
- * \param[in] node         Node action was performed on
- * \param[in] xml_op       XML of operation in CIB status (if available)
- * \param[in] calc_secure  Whether to calculate secure digest
- * \param[in] data_set     Cluster working set
+ * \param[in,out] rsc          Resource that action was for
+ * \param[in]     task         Name of action performed
+ * \param[in]     interval_ms  Action's interval
+ * \param[in]     node         Node action was performed on
+ * \param[in]     xml_op       XML of operation in CIB status (if available)
+ * \param[in]     calc_secure  Whether to calculate secure digest
+ * \param[in,out] data_set     Cluster working set
  *
  * \return Pointer to node's digest cache entry
  */
 static op_digest_cache_t *
 rsc_action_digest(pe_resource_t *rsc, const char *task, guint interval_ms,
-                  pe_node_t *node, xmlNode *xml_op, bool calc_secure,
-                  pe_working_set_t *data_set)
+                  const pe_node_t *node, const xmlNode *xml_op,
```

Comment:
The doxygen here should definitely be [in,out], but const is semantically different and won't always align. We should set const whenever possible to allow the function to be called by a wider range of other functions.

Reply:
Ack. I'm thinking about a specific exchange we had a couple of months ago where I asked this basic question while I was doing some of my own `const`-ification. To the best of my recollection: "How do you feel about using `const` in places where the variable itself doesn't change -- so it is correct from the compiler's perspective -- but something pointed to by the variable does change?" The answer was not to use `const` in such places, basically the same as the current `[in,out]` practice. Unfortunately I can't find the conversation in email, old PRs, or IRC DMs, so all I've got is that vague recollection. Probably in an IRC channel whose history has rolled off. This is why I use email more often now...

I believe the thinking at the time was: It's reasonable for the caller to expect, when passing an argument to a const function, that the argument will emerge totally unchanged when the call is finished. It's also reasonable that this expectation would extend recursively to the argument's members when accessed **via the argument**. Passing a `const` variable to a function that can modify the variable or its members seems like a bug or poor coding practice in most situations, even if the compiler allows it because the "top level" is unmodified.

The person writing the caller can go look for `[in,out]` in the doxygen param specs for the called function, but bugs happen and people forget.

Reply:
My preference isn't very strong so I'm happy to go either way as long as we're consistent.